src/hostapi/pulseaudio/pa_hostapi_pulseaudio.c

```diff
@@ -0,0 +1,1041 @@
+{
+    pa_context_state_t state;
+
+    if (!ptr->context || !ptr->mainloop)
+    {
+        return -1;
+    }
+
+    state = pa_context_get_state(ptr->context);
+
+    if (!PA_CONTEXT_IS_GOOD(state))
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+PaPulseaudioHostApiRepresentation *PulseaudioNew(void)
+{
+    PaPulseaudioHostApiRepresentation *ptr;
+    int fd[2] = { -1, -1 };
+    char proc[PATH_MAX];
+    char buf[PATH_MAX + 20];
+
+    ptr = (PaPulseaudioHostApiRepresentation*)PaUtil_AllocateMemory(sizeof(PaPulseaudioHostApiRepresentation));
+
+    if (!ptr)
+    {
+        return NULL;
+    }
+
+    ptr->mainloop = pa_threaded_mainloop_new();
+
+    if (!ptr->mainloop)
+    {
+        goto fail;
+    }
+
+    memset(buf, 0x00, PATH_MAX + 20);
+    snprintf(buf, sizeof(buf), "Portaudio Pulseaudio HostApi");
+
+    ptr->context =
+        pa_context_new(pa_threaded_mainloop_get_api(ptr->mainloop), buf);
+
+    if (!ptr->context)
+    {
+        goto fail;
+    }
+
+    pa_context_set_state_callback(ptr->context, PulseaudioCheckContextStateCb, ptr);
+
+
+    if (pa_threaded_mainloop_start(ptr->mainloop) < 0)
+    {
+        goto fail;
+    }
+
+    ptr->deviceCount = 0;
+
+    return ptr;
+
+fail:
+    PulseaudioFree(ptr);
+    free(ptr);
+    return NULL;
+}
+
+void PulseaudioFree(PaPulseaudioHostApiRepresentation *ptr)
+{
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_stop(ptr->mainloop);
+    }
+
+    if (ptr->context)
+    {
+        pa_context_disconnect(ptr->context);
+        pa_context_unref(ptr->context);
+    }
+
+    if (ptr->mainloop)
+    {
+        pa_threaded_mainloop_free(ptr->mainloop);
+    }
+
+}
+
+
+static void PulseaudioCheckContextStateCb(pa_context * c, void *userdata)
+{
+    PaPulseaudioHostApiRepresentation *ptr = userdata;
+    assert(c);
+
+    ptr->state = pa_context_get_state(c);
+    pa_threaded_mainloop_signal(ptr->mainloop, 0);
+}
+
+
+void PulseaudioSinkListCb(pa_context *c, const pa_sink_info *l, int eol, void *userdata)
+{
+    PaPulseaudioHostApiRepresentation *l_ptrHostApi = (PaPulseaudioHostApiRepresentation *)userdata;
+    PaError result = paNoError;
+    char *l_strName = NULL;
+
+    /* If eol is set to a positive number, you're at the end of the list */
+    if (eol > 0)
+    {
+        goto error;
+    }
+
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].structVersion = 2;
+    l_ptrHostApi->deviceInfoArray[l_ptrHostApi->deviceCount].hostApi = l_ptrHostApi->hostApiIndex;
+
+    l_strName = (char *)l->name;
+
+    if(l->description != NULL)
+    {
+        l_strName = (char *)l->description;
+    }
+
+    PA_UNLESS(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount] = (char*)PaUtil_GroupAllocateMemory(l_ptrHostApi->allocations,
+              strnlen(l->name, 1024) + 1), paInsufficientMemory);
+
+    memset(l_ptrHostApi->pulseaudioDeviceNames[l_ptrHostApi->deviceCount], 0x00, strnlen(l->name, 1024) + 1);
```

Comment:
Currently it is like this: 

name = "hallo\0"
strnlen(name,3) = 3; 
malloc 3 +1 -> "xxxx"
memset 4 -> "\0\0\0\0"   <- Loop overt the string 
strncpy 3 -> "hal\0"


You can replace it with the following to save the memset call 

name = "hallo\0"
strnlen(name,3) = 3; 
malloc 3 +1 -> "xxxx"
[3] = '\0' "xxx\0"  <- write only one byte 
strncpy 3 -> "hal\0"